File: containers/elastic/elastic__connectors/parquet/python/src/handler.py
import asyncio
import json
import sqlite3
from collections import defaultdict
from contextlib import AsyncExitStack
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Optional, Tuple, Union
import pyarrow
import pyarrow.parquet as pq
import pyarrow.dataset as ds
from prometheus_client import Counter, Gauge, Histogram
import structlog
from tenacity import retry, stop_after_attempt, wait_exponential
from schema import Schema, Schemas
from secrets import token_hex
import aiofiles
import watchfiles
import os
import psutil

logger = structlog.get_logger()

EVENTS_PROCESSED = Counter("parquet_connector_events_processed_total", "Total events processed per file", ["file", "schema"])
ERRORS_TOTAL = Counter("parquet_connector_errors_total", "Total errors per file", ["file", "schema"])
PROCESSING_LATENCY = Histogram("parquet_connector_processing_latency_seconds", "Processing latency per file", ["file", "schema"])
ACTIVE_FILES = Gauge("parquet_connector_active_files", "Active Parquet files being processed", ["schema"])

class HandlerError(Exception):
    pass

class Position:
    def __init__(self, kind: str, value: Union[datetime, str, int]):
        self.kind = kind
        self.value = value

    def to_dict(self) -> Dict:
        return {"kind": self.kind, "value": str(self.value) if isinstance(self.value, datetime) else self.value}

    @classmethod
    def from_dict(cls, data: Dict) -> "Position":
        kind = data["kind"]
        value = data["value"]
        if kind == "Timestamp":
            value = datetime.fromisoformat(value)
        elif kind == "Offset":
            value = int(value)
        return cls(kind, value)

class ParquetHandler:
    def __init__(
        self,
        data_dir: str,
        sender_queue: asyncio.Queue,
        schemas: Schemas,
        batch_size: int,
        poll_interval: float,
        state_path: str,
        incremental_enabled: bool,
        max_files_concurrent: int,
        max_memory_mb: int,
        sqlcipher_key: str = token_hex(16),
    ):
        if not (1 <= batch_size <= 1000):
            raise HandlerError("batch_size must be between 1 and 1000")
        if not (1 <= poll_interval <= 3600):
            raise HandlerError("poll_interval must be between 1 and 3600 seconds")
        if not (1 <= max_files_concurrent <= 50):
            raise HandlerError("max_files_concurrent must be between 1 and 50")
        if not (100 <= max_memory_mb <= 8192):
            raise HandlerError("max_memory_mb must be between 100 and 8192 MB")

        self.data_dir = Path(data_dir).resolve()
        if not self.data_dir.is_dir():
            raise HandlerError(f"Data directory {data_dir} does not exist")
        if not os.access(self.data_dir, os.R_OK):
            raise HandlerError(f"No read permission for {data_dir}")

        self.sender_queue = sender_queue
        self.schemas = schemas
        self.batch_size = batch_size
        self.poll_interval = poll_interval
        self.state_path = state_path
        self.incremental_enabled = incremental_enabled
        self.max_files_concurrent = max_files_concurrent
        self.max_memory_mb = max_memory_mb
        self.sqlcipher_key = sqlcipher_key
        self.active_files: set = set()
        self.last_positions: Dict[str, Position] = self._load_state()

    def _load_state(self) -> Dict[str, Position]:
        try:
            with sqlite3.connect(f"file:{self.state_path}?mode=rw", uri=True) as conn:
                conn.execute(f"PRAGMA key='{self.sqlcipher_key}'")
                conn.execute("CREATE TABLE IF NOT EXISTS positions (file_name TEXT PRIMARY KEY, position TEXT)")
                cursor = conn.cursor()
                cursor.execute("SELECT file_name, position FROM positions")
                return {row[0]: Position.from_dict(json.loads(row[1])) for row in cursor.fetchall()}
        except sqlite3.Error as e:
            logger.error(f"State load error: {e}", exc_info=True)
            return {}

    def _save_state(self, file_name: str, position: Position):
        try:
            with sqlite3.connect(f"file:{self.state_path}?mode=rw", uri=True) as conn:
                conn.execute(f"PRAGMA key='{self.sqlcipher_key}'")
                conn.execute("INSERT OR REPLACE INTO positions (file_name, position) VALUES (?, ?)",
                             (file_name, json.dumps(position.to_dict())))
                conn.commit()
        except sqlite3.Error as e:
            logger.error(f"State save error: {e}", exc_info=True)

    async def start(self):
        if self.incremental_enabled:
            await self._start_incremental()
        else:
            await self._start_full_scan()

    async def _discover_files(self) -> List[Path]:
        logger.info("Discovering Parquet files")
        try:
            start = datetime.now()
            files = [f for f in self.data_dir.glob("*.parquet") if os.access(f, os.R_OK)]
            PROCESSING_LATENCY.labels("discover", "global").observe((datetime.now() - start).total_seconds())
            logger.debug(f"Discovered {len(files)} Parquet files")
            return files
        except Exception as e:
            logger.error(f"File discovery error: {e}", exc_info=True)
            raise HandlerError(f"Failed to discover files: {e}")

    async def _start_full_scan(self):
        while True:
            try:
                files = await self._discover_files()
                for file in files:
                    if len(self.active_files) >= self.max_files_concurrent:
                        await asyncio.sleep(0.1)
                        continue
                    try:
                        await self._process_file(file)
                    except Exception as e:
                        schema_name = self.schemas.get_schema(file.name).name if self.schemas.get_schema(file.name) else "unknown"
                        ERRORS_TOTAL.labels(file.name, schema_name).inc()
                        logger.error(f"Error processing file {file}: {e}", exc_info=True)
            except Exception as e:
                ERRORS_TOTAL.labels("discover", "global").inc()
                logger.error(f"Error discovering files: {e}", exc_info=True)
            await asyncio.sleep(self.poll_interval)

    async def _start_incremental(self):
        async for changes in watchfiles.awatch(self.data_dir, watch_filter=watchfiles.filters.RegexFilter(r".*\.parquet$")):
            try:
                files = await self._discover_files()
                for file in files:
                    if len(self.active_files) >= self.max_files_concurrent:
                        await asyncio.sleep(0.1)
                        continue
                    try:
                        await self._process_file(file)
                    except Exception as e:
                        schema_name = self.schemas.get_schema(file.name).name if self.schemas.get_schema(file.name) else "unknown"
                        ERRORS_TOTAL.labels(file.name, schema_name).inc()
                        logger.error(f"Error processing file {file}: {e}", exc_info=True)
            except Exception as e:
                ERRORS_TOTAL.labels("discover", "global").inc()
                logger.error(f"Error discovering files: {e}", exc_info=True)

    async def _process_file(self, file: Path):
        file_name = file.name
        if file_name in self.active_files:
            return
        self.active_files.add(file_name)
        schema = self.schemas.get_schema(file_name)
        schema_name = schema.name if schema else "unknown"
        ACTIVE_FILES.labels(schema_name).inc()

        if not schema:
            logger.warning(f"No schema found for file {file_name}")
            self.active_files.remove(file_name)
            ACTIVE_FILES.labels(schema_name).dec()
            return

        logger.info(f"Processing file: {file_name}, schema: {schema_name}")
        async with AsyncExitStack() as stack:
            try:
                start = datetime.now()
                # Check memory constraints
                mem_available = psutil.virtual_memory().available // (1024 ** 2)  # MB
                if mem_available < self.max_memory_mb:
                    raise HandlerError(f"Insufficient memory: {mem_available}MB available, {self.max_memory_mb}MB required")

                # Try as single Parquet file, then as dataset
                try:
                    parquet_file = pq.ParquetFile(file)
                    schema_arrow = parquet_file.schema_arrow
                    is_dataset = False
                except pyarrow.ArrowInvalid:
                    dataset = ds.dataset(file, format="parquet")
                    schema_arrow = dataset.schema
                    is_dataset = True

                missing_fields = self._validate_schema(schema_arrow, schema)
                if missing_fields:
                    raise HandlerError(f"Schema mismatch for file {file_name}: missing fields {missing_fields}")

                pos = self.last_positions.get(file_name)
                start_row = 0 if not pos or pos.kind != "Offset" else pos.value
                new_pos = pos

                batch = []
                if is_dataset:
                    # Process partitioned dataset
                    for fragment in dataset.get_fragments():
                        for batch_record in fragment.to_table().to_batches():
                            df = batch_record.to_pandas()
                            for _, row in df.iterrows():
                                if start_row > 0:
                                    start_row -= 1
                                    continue
                                event = row.to_dict()
                                if self.incremental_enabled and schema.timestamp_field:
                                    ts = event.get(schema.timestamp_field)
                                    if ts and pos and pos.kind == "Timestamp" and ts <= pos.value:
                                        continue
                                transformed = self._transform_to_ecs(event, schema)
                                batch.append(transformed)
                                EVENTS_PROCESSED.labels(file_name, schema_name).inc()

                                if schema.timestamp_field and (ts := event.get(schema.timestamp_field)):
                                    new_pos = Position("Timestamp", ts)
                                else:
                                    new_pos = Position("Offset", start_row + len(df))

                                if len(batch) >= self.batch_size:
                                    await self.sender_queue.put((batch, file_name, len(batch)))
                                    batch = []
                else:
                    # Process single Parquet file
                    num_rows = parquet_file.metadata.num_rows
                    for batch_idx in range(parquet_file.num_row_groups):
                        if start_row >= num_rows:
                            break
                        table = parquet_file.read_row_group(batch_idx, use_memory_map=True)
                        df = table.to_pandas()
                        for _, row in df.iterrows():
                            if start_row > 0:
                                start_row -= 1
                                continue
                            event = row.to_dict()
                            if self.incremental_enabled and schema.timestamp_field:
                                ts = event.get(schema.timestamp_field)
                                if ts and pos and pos.kind == "Timestamp" and ts <= pos.value:
                                    continue
                            transformed = self._transform_to_ecs(event, schema)
                            batch.append(transformed)
                            EVENTS_PROCESSED.labels(file_name, schema_name).inc()

                            if schema.timestamp_field and (ts := event.get(schema.timestamp_field)):
                                new_pos = Position("Timestamp", ts)
                            else:
                                new_pos = Position("Offset", start_row + len(df))

                            if len(batch) >= self.batch_size:
                                await self.sender_queue.put((batch, file_name, len(batch)))
                                batch = []

                if batch:
                    await self.sender_queue.put((batch, file_name, len(batch)))

                if new_pos:
                    self.last_positions[file_name] = new_pos
                    self._save_state(file_name, new_pos)

                PROCESSING_LATENCY.labels(file_name, schema_name).observe((datetime.now() - start).total_seconds())
            except pyarrow.ArrowInvalid as e:
                ERRORS_TOTAL.labels(file_name, schema_name).inc()
                logger.error(f"Invalid Parquet file {file_name}: {e}", exc_info=True)
                raise HandlerError(f"Failed to process file {file_name}: {e}")
            except OSError as e:
                ERRORS_TOTAL.labels(file_name, schema_name).inc()
                logger.error(f"File access error for {file_name}: {e}", exc_info=True)
                raise HandlerError(f"Failed to process file {file_name}: {e}")
            except Exception as e:
                ERRORS_TOTAL.labels(file_name, schema_name).inc()
                logger.error(f"File processing error for {file_name}: {e}", exc_info=True)
                raise HandlerError(f"Failed to process file {file_name}: {e}")
            finally:
                self.active_files.remove(file_name)
                ACTIVE_FILES.labels(schema_name).dec()

    def _validate_schema(self, arrow_schema, schema: Schema) -> List[str]:
        expected_fields = set(v for k, v in schema.mappings.ecs.items() if k != "sourcetype" and not v.startswith('"'))
        actual_fields = set(arrow_schema.names)
        missing_fields = expected_fields - actual_fields
        if missing_fields:
            logger.warning(f"Schema validation failed: missing fields {missing_fields} in {schema.file_name}")
        return list(missing_fields)

    def _transform_to_ecs(self, event: Dict, schema: Schema) -> Dict:
        ecs = {
            "@timestamp": event.get(schema.mappings.ecs.get("@timestamp", "timestamp"), datetime.utcnow().isoformat()),
            "ecs": {"version": "8.0.0"},
        }
        for key, value in schema.mappings.ecs.items():
            if not key.startswith("@"):
                if value.startswith('"') and value.endswith('"'):
                    ecs[key] = value.strip('"')
                else:
                    ecs[key] = event.get(value, "")
        return ecs

File: containers/elastic/elastic__connectors/parquet/rust/src/handler.rs
use anyhow::{Context, Result};
use log::{error, info, warn};
use notify::{Config, RecommendedWatcher, RecursiveMode, Watcher};
use parquet::file::reader::{FileReader, SerializedFileReader};
use parquet::arrow::ArrowReader;
use arrow::record_batch::RecordBatch;
use arrow::datatypes::SchemaRef;
use arrow::io::parquet::read::read_metadata;
use arrow::io::parquet::read::ParquetFile;
use prometheus::{Counter, Gauge, Histogram, Registry};
use rusqlite::{params, Connection};
use serde_json::Value;
use std::collections::HashSet;
use std::path::Path;
use std::sync::Arc;
use std::time::{Duration, SystemTime};
use tokio::sync::mpsc;
use tokio::time;

use crate::schema::{Schema, Schemas};

lazy_static::lazy_static! {
    static ref EVENTS_PROCESSED: Counter = Counter::with_opts(prometheus::opts!(
        "parquet_connector_events_processed_total",
        "Total events processed per file",
        &["file", "schema"]
    ))
    .unwrap();
    static ref ERRORS_TOTAL: Counter = Counter::with_opts(prometheus::opts!(
        "parquet_connector_errors_total",
        "Total errors per file",
        &["file", "schema"]
    ))
    .unwrap();
    static ref PROCESSING_LATENCY: Histogram = Histogram::with_opts(
        prometheus::histogram_opts!(
            "parquet_connector_processing_latency_seconds",
            "Processing latency per file",
            &["file", "schema"]
        )
    )
    .unwrap();
    static ref ACTIVE_FILES: Gauge = Gauge::with_opts(prometheus::opts!(
        "parquet_connector_active_files",
        "Active Parquet files being processed",
        &["schema"]
    ))
    .unwrap();
}

#[derive(thiserror::Error, Debug)]
pub enum HandlerError {
    #[error("Invalid configuration: {0}")]
    Config(String),
    #[error("File error: {0}")]
    File(String),
    #[error("Database error: {0}")]
    Database(String),
    #[error("Schema error: {0}")]
    Schema(String),
}

#[derive(serde::Serialize, serde::Deserialize)]
struct Position {
    kind: String,
    value: String,
}

impl Position {
    fn new_timestamp(timestamp: SystemTime) -> Self {
        Self {
            kind: "Timestamp".to_string(),
            value: timestamp
                .duration_since(SystemTime::UNIX_EPOCH)
                .unwrap()
                .as_secs()
                .to_string(),
        }
    }

    fn new_offset(offset: usize) -> Self {
        Self {
            kind: "Offset".to_string(),
            value: offset.to_string(),
        }
    }
}

pub struct ParquetHandler {
    data_dir: String,
    sender_queue: mpsc::Sender<(Vec<Value>, String, usize)>,
    schemas: Arc<Schemas>,
    batch_size: usize,
    poll_interval: f64,
    state_path: String,
    incremental_enabled: bool,
    max_files_concurrent: usize,
    max_memory_mb: usize,
    sqlcipher_key: String,
    active_files: HashSet<String>,
    registry: Arc<Registry>,
}

impl ParquetHandler {
    pub fn new(
        data_dir: String,
        sender_queue: mpsc::Sender<(Vec<Value>, String, usize)>,
        schemas: Arc<Schemas>,
        batch_size: usize,
        poll_interval: f64,
        state_path: String,
        incremental_enabled: bool,
        max_files_concurrent: usize,
        max_memory_mb: usize,
        sqlcipher_key: String,
        registry: Arc<Registry>,
    ) -> Result<Self> {
        if !(1..=1000).contains(&batch_size) {
            return Err(HandlerError::Config("batch_size must be between 1 and 1000".to_string()).into());
        }
        if !(1.0..=3600.0).contains(&poll_interval) {
            return Err(HandlerError::Config("poll_interval must be between 1 and 3600 seconds".to_string()).into());
        }
        if !(1..=50).contains(&max_files_concurrent) {
            return Err(HandlerError::Config("max_files_concurrent must be between 1 and 50".to_string()).into());
        }
        if !(100..=8192).contains(&max_memory_mb) {
            return Err(HandlerError::Config("max_memory_mb must be between 100 and 8192 MB".to_string()).into());
        }

        let path = Path::new(&data_dir);
        if !path.is_dir() {
            return Err(HandlerError::File(format!("Data directory {} does not exist", data_dir)).into());
        }
        if !std::fs::metadata(&data_dir)
            .map(|m| m.permissions().readonly())
            .unwrap_or(true)
        {
            return Err(HandlerError::File(format!("No read permission for {}", data_dir)).into());
        }

        Ok(Self {
            data_dir,
            sender_queue,
            schemas,
            batch_size,
            poll_interval,
            state_path,
            incremental_enabled,
            max_files_concurrent,
            max_memory_mb,
            sqlcipher_key,
            active_files: HashSet::new(),
            registry,
        })
    }

    pub async fn start(&self) -> Result<()> {
        self.registry.register(Box::new(EVENTS_PROCESSED.clone()))?;
        self.registry.register(Box::new(ERRORS_TOTAL.clone()))?;
        self.registry.register(Box::new(PROCESSING_LATENCY.clone()))?;
        self.registry.register(Box::new(ACTIVE_FILES.clone()))?;

        if self.incremental_enabled {
            self.start_incremental().await
        } else {
            self.start_full_scan().await
        }
    }

    async fn start_full_scan(&self) -> Result<()> {
        let mut interval = time::interval(Duration::from_secs_f64(self.poll_interval));
        loop {
            interval.tick().await;
            if let Err(e) = self.process_files().await {
                error!("Error discovering files: {}", e);
                ERRORS_TOTAL.with_label_values(&["discover", "global"]).inc();
            }
        }
    }

    async fn start_incremental(&self) -> Result<()> {
        let mut watcher = RecommendedWatcher::new(
            move |res| {
                if let Err(e) = res {
                    error!("File watch error: {}", e);
                }
            },
            Config::default(),
        )?;
        watcher.watch(Path::new(&self.data_dir), RecursiveMode::NonRecursive)?;
        loop {
            if let Err(e) = self.process_files().await {
                error!("Error discovering files: {}", e);
                ERRORS_TOTAL.with_label_values(&["discover", "global"]).inc();
            }
            tokio::time::sleep(Duration::from_secs_f64(self.poll_interval)).await;
        }
    }

    async fn process_files(&self) -> Result<()> {
        let files: Vec<_> = std::fs::read_dir(&self.data_dir)?
            .filter_map(|entry| {
                let path = entry.ok()?.path();
                if path.extension().map(|ext| ext == "parquet").unwrap_or(false) {
                    Some(path.to_str()?.to_string())
                } else {
                    None
                }
            })
            .collect();
        info!("Discovered {} Parquet files", files.len());

        for file in files {
            if self.active_files.len() >= self.max_files_concurrent {
                tokio::time::sleep(Duration::from_millis(100)).await;
                continue;
            }
            if let Err(e) = self.process_file(&file).await {
                let file_name = Path::new(&file).file_name().unwrap().to_str().unwrap();
                let schema_name = self.schemas.get_schema(file_name).map(|s| s.name).unwrap_or("unknown".to_string());
                ERRORS_TOTAL.with_label_values(&[file_name, &schema_name]).inc();
                error!("Error processing file {}: {}", file, e);
            }
        }
        Ok(())
    }

    async fn process_file(&self, file: &str) -> Result<()> {
        let file_name = Path::new(file).file_name().unwrap().to_str().unwrap().to_string();
        if self.active_files.contains(&file_name) {
            return Ok(());
        }
        self.active_files.insert(file_name.clone());
        let schema = self.schemas.get_schema(&file_name);
        let schema_name = schema.as_ref().map(|s| s.name.clone()).unwrap_or("unknown".to_string());
        ACTIVE_FILES.with_label_values(&[&schema_name]).inc();

        if schema.is_none() {
            warn!("No schema found for file {}", file_name);
            self.active_files.remove(&file_name);
            ACTIVE_FILES.with_label_values(&[&schema_name]).dec();
            return Ok(());
        }
        let schema = schema.unwrap();
        info!("Processing file: {}, schema: {}", file_name, schema_name);

        let conn = Connection::open(&self.state_path)?;
        conn.pragma_update(None, "key", &self.sqlcipher_key)?;
        conn.execute(
            "CREATE TABLE IF NOT EXISTS positions (file_name TEXT PRIMARY KEY, position TEXT)",
            [],
        )?;

        let mut batch = Vec::new();
        let pos = self.load_position(&conn, &file_name)?;
        let start_row = if pos.kind == "Offset" {
            pos.value.parse::<usize>().unwrap_or(0)
        } else {
            0
        };
        let mut new_pos = pos.clone();

        // Try as single Parquet file, then as dataset
        let mut is_dataset = false;
        let schema_arrow: SchemaRef;
        let mut record_batches: Box<dyn Iterator<Item = Result<RecordBatch>>>;

        if let Ok(file_reader) = SerializedFileReader::new(std::fs::File::open(file)?) {
            let arrow_reader = ArrowReader::new(file_reader);
            schema_arrow = arrow_reader.get_schema()?;
            record_batches = Box::new(arrow_reader);
        } else {
            let dataset = arrow::dataset::ParquetFile::try_new(file)?;
            schema_arrow = dataset.schema().clone();
            record_batches = Box::new(dataset.into_iter());
            is_dataset = true;
        }

        let missing_fields = self.validate_schema(&schema_arrow, &schema);
        if !missing_fields.is_empty() {
            return Err(HandlerError::Schema(format!("Missing fields: {:?}", missing_fields)).into());
        }

        for record_batch in record_batches {
            let record_batch = record_batch?;
            for row in 0..record_batch.num_rows() {
                if row < start_row {
                    continue;
                }
                let event = self.row_to_event(&record_batch, row)?;
                if self.incremental_enabled && schema.timestamp_field.is_some() {
                    let ts = event
                        .get(schema.timestamp_field.as_ref().unwrap())
                        .and_then(|v| v.as_i64())
                        .map(|v| SystemTime::UNIX_EPOCH + Duration::from_secs(v as u64));
                    if let (Some(ts), Position { kind, value, .. }) = (ts, &pos) {
                        if kind == "Timestamp" && ts <= (SystemTime::UNIX_EPOCH + Duration::from_secs(value.parse::<u64>().unwrap_or(0))) {
                            continue;
                        }
                    }
                }
                let transformed = self.transform_to_ecs(&event, &schema);
                batch.push(transformed);
                EVENTS_PROCESSED.with_label_values(&[&file_name, &schema_name]).inc();

                if schema.timestamp_field.is_some() {
                    if let Some(ts) = event.get(schema.timestamp_field.as_ref().unwrap()).and_then(|v| v.as_i64()) {
                        new_pos = Position::new_timestamp(SystemTime::UNIX_EPOCH + Duration::from_secs(ts as u64));
                    }
                } else {
                    new_pos = Position::new_offset(row + 1);
                }

                if batch.len() >= self.batch_size {
                    self.sender_queue
                        .send((batch.clone(), file_name.clone(), batch.len()))
                        .await
                        .context("Failed to send batch")?;
                    batch.clear();
                }
            }
        }

        if !batch.is_empty() {
            self.sender_queue
                .send((batch.clone(), file_name.clone(), batch.len()))
                .await
                .context("Failed to send batch")?;
        }

        if new_pos != pos {
            self.save_position(&conn, &file_name, &new_pos)?;
        }

        self.active_files.remove(&file_name);
        ACTIVE_FILES.with_label_values(&[&schema_name]).dec();
        Ok(())
    }

    fn load_position(&self, conn: &Connection, file_name: &str) -> Result<Position> {
        let mut stmt = conn.prepare("SELECT position FROM positions WHERE file_name = ?")?;
        let pos: Option<String> = stmt
            .query_row(params![file_name], |row| row.get(0))
            .optional()?;
        Ok(pos
            .map(|p| serde_json::from_str(&p).context("Failed to parse position"))
            .transpose()?
            .unwrap_or(Position {
                kind: "Offset".to_string(),
                value: "0".to_string(),
            }))
    }

    fn save_position(&self, conn: &Connection, file_name: &str, position: &Position) -> Result<()> {
        let pos_json = serde_json::to_string(position)?;
        conn.execute(
            "INSERT OR REPLACE INTO positions (file_name, position) VALUES (?, ?)",
            params![file_name, pos_json],
        )?;
        Ok(())
    }

    fn validate_schema(&self, arrow_schema: &SchemaRef, schema: &Schema) -> Vec<String> {
        let expected_fields: HashSet<_> = schema
            .mappings
            .ecs
            .iter()
            .filter(|(k, _)| k != &"sourcetype")
            .filter(|(_, v)| !v.starts_with('"') || !v.ends_with('"'))
            .map(|(_, v)| v.clone())
            .collect();
        let actual_fields: HashSet<_> = arrow_schema.fields().iter().map(|f| f.name().clone()).collect();
        expected_fields.difference(&actual_fields).cloned().collect()
    }

    fn row_to_event(&self, batch: &RecordBatch, row: usize) -> Result<Value> {
        let mut event = serde_json::Map::new();
        for (i, column) in batch.columns().iter().enumerate() {
            let name = batch.schema().field(i).name();
            let value = column.get(row)?;
            event.insert(name.to_string(), value.into());
        }
        Ok(Value::Object(event))
    }

    fn transform_to_ecs(&self, event: &Value, schema: &Schema) -> Value {
        let mut ecs = serde_json::Map::new();
        ecs.insert(
            "@timestamp".to_string(),
            event
                .get(schema.mappings.ecs.get("@timestamp").unwrap_or(&"timestamp".to_string()))
                .cloned()
                .unwrap_or(Value::String(chrono::Utc::now().to_rfc3339())),
        );
        ecs.insert("ecs".to_string(), Value::Object(serde_json::Map::from_iter(vec![(
            "version".to_string(),
            Value::String("8.0.0".to_string()),
        )])));

        for (key, value) in &schema.mappings.ecs {
            if !key.starts_with('@') {
                if value.starts_with('"') && value.ends_with('"') {
                    ecs.insert(key.clone(), Value::String(value.trim_matches('"').to_string()));
                } else {
                    ecs.insert(key.clone(), event.get(value).cloned().unwrap_or(Value::String("".to_string())));
                }
            }
        }

        Value::Object(ecs)
    }
}